mysqlnd_azure.c

```diff
@@ -157,14 +157,183 @@ set_redirect_client_options(MYSQLND_CONN_DATA * const conn, MYSQLND_CONN_DATA *
+		return FALSE;
+	}
+
+	char  redirect_port[8] = { 0 };
+	memcpy(redirect_port, port_begin, port_len);
+	int port = atoi(redirect_port);
+	if (port <= 0) {
+		return FALSE;
+	}
+
+	if (ttl_len > 0) {
+		char  redirect_ttl[8] = { 0 };
+		memcpy(redirect_ttl, ttl_begin, ttl_len);
+		int ttl = atoi(redirect_ttl);
+		if (ttl <= 0 && strcmp(redirect_ttl, "0") != 0) {
+			return FALSE;
+		}
+		*p_ui_redirect_ttl = ttl;
+	}
+
+	//setback the value when everything is settled
+	*p_ui_redirect_port = port;
+	memcpy(redirect_host, host_begin, host_len);
+	memcpy(redirect_user, user_begin, user_len);
+
+	return TRUE;
+}
+
+static zend_bool
+parse_community_protocol(const MYSQLND_STRING * const last_message, char* redirect_host, char* redirect_user, unsigned int* p_ui_redirect_port, unsigned int* p_ui_redirect_ttl)
+{
+	/**
+	* Community protocol:
+	* Location: mysql://[redirectedHostName]:redirectedPort/?user=redirectedUser&ttl=%d/n
+	*/
+	const char* msg_header = "Location: mysql://[";
+	int msg_header_len = strlen(msg_header);
+	char* cur_pos = last_message->s + msg_header_len;
+	char* end = last_message->s + last_message->l;
+
+	char* host_begin = cur_pos, * host_end = NULL,
+		* port_begin = NULL, * port_end = NULL,
+		* user_begin = NULL, * user_end = NULL,
+		* ttl_begin = NULL, * ttl_end = NULL;
+
+	host_end = strchr(cur_pos, ']');
+	if (host_end == NULL) return FALSE;
+
+	cur_pos = host_end + 1;
+	if (cur_pos == end || *cur_pos != ':' || ++cur_pos == end) return FALSE;
+
+	port_begin = cur_pos;
+	port_end = strchr(cur_pos, '/');
+	if (port_end == NULL) return FALSE;
+
+	cur_pos = port_end + 1;
+	if (cur_pos == end || *cur_pos != '?' || ++cur_pos == end) return FALSE;
+
+	int user_delimiter_len = strlen("user=");
+	if (end - cur_pos <= user_delimiter_len || strncmp(cur_pos, "user=", user_delimiter_len) != 0) return FALSE;
+
+	user_begin = cur_pos + user_delimiter_len;
+	user_end = strchr(cur_pos, '&');
+	if (user_end == NULL) return FALSE;
+
+	cur_pos = user_end + 1;
+	if (cur_pos == end) return FALSE;
+
+	int ttl_delimiter_len = strlen("ttl=");
+	if (end - cur_pos <= ttl_delimiter_len || strncmp(cur_pos, "ttl=", ttl_delimiter_len) != 0) return FALSE;
+
+	ttl_begin = cur_pos + ttl_delimiter_len;
+	ttl_end = strchr(cur_pos, '\n');
+	if (ttl_end == NULL) return FALSE;
+
+	int host_len = host_end - host_begin;
+	int port_len = port_end - port_begin;
+	int user_len = user_end - user_begin;
+	int ttl_len = ttl_end - ttl_begin;
+
+	if (host_len <= 0 || port_len <= 0 || user_len <= 0 || ttl_len <= 0 || host_len > MAX_REDIRECT_HOST_LEN || port_len > 8 || user_len > MAX_REDIRECT_USER_LEN || ttl_len > 8) {
+		return FALSE;
+	}
+
+	char  redirect_port[8] = { 0 };
+	memcpy(redirect_port, port_begin, port_len);
+	int port = atoi(redirect_port);
+	if (port <= 0) {
+		return FALSE;
+	}
+
+	char  redirect_ttl[8] = { 0 };
+	memcpy(redirect_ttl, ttl_begin, ttl_len);
+	int ttl = atoi(redirect_ttl);
+	if (ttl <= 0 && strcmp(redirect_ttl, "0") != 0) {
```

Comment:
Per my understanding the strcmp here is to make sure the return value 0 from atoi is a product of parsing "0", not a signal of parsing error. However, quoting C++ reference:

If the converted value would be out of the range of representable values by an int, it causes undefined behavior. See strtol for a more robust cross-platform alternative when this is a possibility.

So, would you consider using strtol(or strtoul) to get a deterministic behaviour?

Reply:
Thanks for your suggestion. I've updated the code. Can you help check? Thanks.